### ISSUE TITLE
Scope of sos_tempfile target.

### ISSUE THREAD

**Issue by BoPeng** [vatlab/sos#1369]
@N7DR raised questions in gitter as how to transfer "output" from one step to another one, without having to create a file. In particular, his definition of "output" can be "output" written to the terminal from programs.

There are two related features:

1. **disk-based** Capturing output to a file and pass it to another step. For example, in the following example the `stdout` was captured to `a.txt`, which is the output of step `a`, and is passed to step `default` as input.

```
[a]
output: 'a.txt'
sh: stdout='a.txt'
  echo "whatever"

[default]
input: 'a.txt'
sh: expand=True
   cat {_input}
```

2. **memory-based** Sending variables across through variables,

```
[a: shared='out_var']

sh: stdout='a.txt'
  echo "whatever"

out_var = open('a.txt').read()

[default]
depends: sos_variable('out_var')
sh: expand=True
   echo {out_var}
```

What was requested was some sort of "unnamed" output, something like:
```
[a]

sh:
  echo "whatever"

[default]
out_a = sos_output('a')
sh: expand=True
   echo {out_a}
```
Basically holding output from step 'a' to a variable that can be retrieved from other steps (and creating a dependency because of this.)

Depending on the size of output, there can be disk based and memory based solutions. This ticket is created to see if anyone else is interested in such a feature, and for discussions on interface and implementation of such a feature.

**Comment by BoPeng** [comment]
I think one problem with the first (disk-based) scenario is that people run out of names for temporary files and would like to remove these intermediate files automatically.

So instead of

```
[a]
output: 'a.txt'
sh: stdout='a.txt'
  echo "whatever"

[default]
input: 'a.txt'
sh: expand=True
   cat {_input}
```
we could have some named temp file, such as

```
[a]
output: sos_temp('a')
sh: stdout=_output
  echo "whatever"

[default]
input: output_from('a')
sh: expand=True
   cat {_input}
```

It is a small change but `sos_temp` would create a temporary file, which would be automatically removed after `sos` terminates.

**Comment by BoPeng** [comment]
The memory-based scenario could be improved from

```
[a: shared='out_var']

sh: stdout='a.txt'
  echo "whatever"

out_var = open('a.txt').read()

[default]
depends: sos_variable('out_var')
sh: expand=True
   echo {out_var}
```
to
```
[a: shared='out_var']

sh: stdout=sos_variable('out_var')
  echo "whatever"


[default]
depends: sos_variable('out_var')
sh: expand=True
   echo {out_var}
```

**Comment by N7DR** [comment]
Bo wrote on 4/29/20 8:50 PM:
> I think one problem with the first (disk-based) scenario is that people run out of names for temporary files and would like to remove these intermediate files automatically.

Yes, I probably communicated it poorly, but part of my objection to using
temporary files as things are currently implemented is roughly:

1. You have to think of a name that is guaranteed both to be unique;

2. You need to be sure that the file is still around when you read it in
another step (i.e., it isn't prematurely deleted);

3. You need to get rid of the file when the sequence of steps is complete
(otherwise you can end up with lots of temporary files hanging around ... I
hate with a passion programs that litter /tmp with useless leftover files).

All this could be managed even with the current SoS workflow code, but it's an
ugly mess for the user to do himself, and, if SoS ends up insisting on using
files, it should handle all the bookkeeping itself automagically, so that the
user doesn't have to worry about any of this stuff.

I'm a great believer in having computers do all the error-prone nasty bits of
bookkeeping (like remembering temporary filenames) so that users can
concentrate on the real problem they're trying to solve.

> we could have some named temp file, such as
> 
> ```
> [a]
> output: sos_temp('a')
> sh: stdout=_output
>   echo "whatever"
> 
> [default]
> input: output_from('a')
> sh: expand=True
>    cat {_input}
> ```
> 
> It is a small change but `sos_temp` would create a temporary file, which would be automatically removed after `sos` terminates.
> 

That certainly looks like a great improvement in usability, if one has to use
files.

  Doc

PS I looked at the particular use-case where I first ran into wanting to use
stdout from one step in another step, and the output I was trying to transfer
is just a short line of text, maybe 25 characters (although it can take
several hours of CPU time to generate that line :-) )... so having to jump
through hoops just to pass ~25 bytes definitely seemed like something that
should be made much easier.

-- 
Web:  http://enginehousebooks.com/drevans

**Comment by N7DR** [comment]
Bo wrote on 4/29/20 9:43 PM:
> The memory-based scenario could be improved from
> 
> ```
> [a: shared='out_var']
> 
> sh: stdout='a.txt'
>   echo "whatever"
> 
> out_var = open('a.txt').read()
> 
> [default]
> depends: sos_variable('out_var')
> sh: expand=True
>    echo {out_var}
> ```
> to
> ```
> [a: shared='out_var']
> 
> sh: stdout=sos_variable('out_var')
>   echo "whatever"
> 
> 
> [default]
> depends: sos_variable('out_var')
> sh: expand=True
>    echo {out_var}
> ```
> 

YES! I came up with something precisely equivalent to this as I was pondering
the problem during the night last night :-)

  Doc

-- 
Web:  http://enginehousebooks.com/drevans

**Comment by BoPeng** [comment]
Thanks for the feedback, I think both improvements make sense but I am pinging @gaow for a second opinion.

**Comment by gaow** [comment]
In snakemake there is a `temp()` function for temporary files. But one will still have to come up with a file name: 

https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#protected-and-temporary-files

I dont think coming up with a file name is too much of a trouble, eg, one option is to base the filename on input like `f'{_input:n}.tmp'`.  

I personally prefer to always use input and output files -- pipelines are more robust that way. What happens if we pass variables around for the scenario @N7DR mentioned: "the output I was trying to transfer is just a short line of text, maybe 25 characters (although it can take several hours of CPU time to generate that line :-) ).". Without a file output, when we rerun the pipeline it will have to redo hours of computation, or we still can somehow cache it, eg in `.sos` folder? 

I think caching the stdout make more sense. Then behavior of tempfile and passing stdout will be different. Tempfile are meant to be removed after the workflow completes and the computation will have to be done all over again in a rerun. But stdout will be cached and will not trigger rerun unless `.sos` folder is removed.

I'm cool with the proposed interface for both applications. I suppose `sos_temp()` can take any string input parameter to create as many temp files as a step needs? Then it does not have to be text file but rather any file format, and will be removed as soon as steps depending on these files are complete? 

`stdout=sos_variable()` interface looks good. A more general question is, should we cache `sos_variable` regardless whether it is for `stdout` or otherwise obtained?

**Comment by BoPeng** [comment]
I tend to over-complicated things... 

So in the `sos_tempfile` case, my draft implementation here has a few usages, although the baseline is that all `sos_tempfile` will be automatically removed right before sos shuts down. `snakemake` says "after all rules that use it is completed", which is better but a bit more difficult to implement.

1. Scenario one: We just need a temp file, and I do not care what name you use.

```
[a]
output: sos_tempfile()

sh: stdout=_output
  echo "whatever"


[default]
input: output_from('a')

sh: expand=True
   cat {_input}

```
Note that once the `sos_tempfile()` is created, it is referred by `_output` and `output_from('a')` so a name does not matter.

2. Refer to the same tempfile by name.

```
[a]
output: sos_tempfile('echo')

sh: stdout=_output
  echo "whatever"


[default]
input: sos_tempfile('echo')

sh: expand=True
   cat {_input}
```
In this case a name is given to the temp file, which has nothing to do with actual filename, but sos guarantees that the same name refers to the same temp file.

3. If the path actually matters, and you only wants SoS to remove the file after the workflow is completed (the snakemake usage).

```
[a]
output: sos_tempfile(path='a.txt')
sh: stdout=_output
  echo "whatever"

[default]
input: output_from('a')

sh: expand=True
   cat {_input}
```

4. Only `dir`, `prefix` or `suffix` of the temp file matters (as in the parameters of `tempfile.mkstemp`),

```
[a]
output: sos_tempfile(dir='.', suffix='.txt')
sh: stdout=_output
  echo "whatever"

[default]
input: output_from('a')

sh: expand=True
   cat {_input}
```

I felt that this mechanism can be used internally for the removal of temporary files. Right now we try hard to remove temporary files immediately after they are used, but we could use this mechanism to register temp files and let sos handles removal upon completion.

**Comment by gaow** [comment]
Indeed this is getting complicated. I think 1 ~ 3 are well motivated, but not sure about the necessity for 4. 

> the path actually matters, and you only wants SoS to remove the file after the workflow is completed (the snakemake usage).

I think we will remove all temp files after SoS completes anyways? The only motivation for allowing one to set full name is when temp file has to be written to specific paths not system's default `/tmp` or  `/var/tmp`. So to simplify 3, we can allow for `sos_tempfile(path(...))` and if the input is instance of `path` then we treat it a full path?

**Comment by BoPeng** [comment]
Usage 4 is sometimes justified because processing script might look for particular file extension to determine what to do (`suffix='.fastq'` or `suffix='.fastq.gz`), or if the temp file has to be in the workdir (`dir='.'`). I really cannot think of a usage for `prefix`, but I kept it because `tempfile.mkstemp` has it.

If you agree 1-3 are justifiable, let us do this:

1. `sos_tempfile()`, create a unique temp file in system temp directory.
2. `sos_tempfile(dir, suffix, prefix)`, create a unique temp file in specified directory, or with specified suffix, prefix.
3. `sos_tempfile(name='')`, create a named temp file to be referred repeatedly.
4. `sos_tempfile(...)` or `sos_tempfile(path=...)` to be a filename with type `str`), `path`, or `file_target`. This basically tells `sos` the file should be removed later on.

Note that `sos_tempfile` **does not create any file or open any file**, it just provide a filename.

**Comment by gaow** [comment]
Okay!

> A complication is that SoS should not create any signature if the input or output is sos_tempfile.

Will this trigger rerun? Thinking of `zapped` files, a rerun is only triggered if the steps generating those files before they get zapped, or steps using zapped file as input are changed. Does it make sense to do the same for sos_tempfile?

**Comment by BoPeng** [comment]
> Will this trigger rerun? 

Yes, because with `zap` we have a placeholder to hold the file information, and with `sos_tempfile` the actual file is removed so no information can be found.

**Comment by BoPeng** [comment]
At this point I think we can merge these two features.

With `zap` we immediately replace a (large) file with a placeholder file that physically exists, usually in working directory. The information helps SoS check step signature so the step would continue as if the entire file is still there. This is not particular easy to use since users would need to call something like `_output.zap()` explicitly. 

So perhaps using `_output.zap()`, we do something like

```
output: sos_tempfile(whatever)

sos_tempfile(some_intermediate_file)
sos_tempdir(some_intermediate_dir)??
```
which tells sos that these files and directories should be removed afterwards. However, we keep information on these files somewhere so that the signature will still be recorded and be used to avoid rerun of the step.

So we improve `sos_tempfile` by keeping track of their information and use signatures, and we use a concept that is more familiar to users than "zap". 

Note that with this approach `zap` would be done at the end, so it cannot be used for users who want to control the size of project during the execution of the workflow. That said, the syntax is there and SoS can always be improved to remove the files more intelligently.

**Comment by gaow** [comment]
> At this point I think we can merge these two features.

Yes I think this makes sense. 

> so it cannot be used for users who want to control the size of project during the execution of the workflow. 

This will be a major limition to my benchmark applications where I just want to get rid of large steps as soon as possible. However indeed we can think about improving that later if we are enough motivated by applications.

**Comment by BoPeng** [comment]
>  However indeed we can think about improving that later if we are enough motivated by applications.

We are again bitten by the flexibility of SoS in this regard, because it is very difficult to predict if a target will no longer be used, given that a nested workflow could be set up later with something like `sos_run("some_workflow")` when `some_workflow` could be set up dynamically and refer to something created before. 

So in addition to using `zap` explicitly, users would have to inform `sos_tempfile` the scope of tempfile with something like

```
sos_tempfile(scope='workflow')
```
and the tempfile would by default be removed after the completion of a "workflow" that it is created, and then
```
sos_tempfile(scope='step')
```
will remove it as soon as the step is completed, and
```
sos_tempfile(scope='master')
```
will remove at the end of the entire sos run.
